File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/models/cells_windows.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

CellsWindowsTargetKind = Literal["int", "double", "char", "window"]


@dataclass(frozen=True)
class StepCellsWindowsCard:
    op: Literal["cells-windows.card"]
    targetKind: CellsWindowsTargetKind
    targetId: str


@dataclass(frozen=True)
class StepCellsWindowsInsrtc:
    op: Literal["cells-windows.insrtc"]
    cellId: str
    item: str
    maxCardinality: int | None = None
    length: int | None = None


@dataclass(frozen=True)
class StepCellsWindowsInsrtd:
    op: Literal["cells-windows.insrtd"]
    cellId: str
    item: float
    maxCardinality: int | None = None


@dataclass(frozen=True)
class StepCellsWindowsInsrti:
    op: Literal["cells-windows.insrti"]
    cellId: str
    item: int
    maxCardinality: int | None = None


@dataclass(frozen=True)
class StepCellsWindowsScard:
    op: Literal["cells-windows.scard"]
    card: int
    targetKind: CellsWindowsTargetKind
    targetId: str


@dataclass(frozen=True)
class StepCellsWindowsSize:
    op: Literal["cells-windows.size"]
    targetKind: CellsWindowsTargetKind
    targetId: str


@dataclass(frozen=True)
class StepCellsWindowsSsize:
    op: Literal["cells-windows.ssize"]
    size: int
    targetKind: CellsWindowsTargetKind
    targetId: str


@dataclass(frozen=True)
class StepCellsWindowsValid:
    op: Literal["cells-windows.valid"]
    size: int
    n: int
    targetKind: CellsWindowsTargetKind
    targetId: str


@dataclass(frozen=True)
class StepCellsWindowsWncard:
    op: Literal["cells-windows.wncard"]
    windowId: str


@dataclass(frozen=True)
class StepCellsWindowsWninsd:
    op: Literal["cells-windows.wninsd"]
    windowId: str
    left: float
    right: float
    maxIntervals: int | None = None


@dataclass(frozen=True)
class StepCellsWindowsWnfetd:
    op: Literal["cells-windows.wnfetd"]
    windowId: str
    index: int


@dataclass(frozen=True)
class StepCellsWindowsWnvald:
    op: Literal["cells-windows.wnvald"]
    size: int
    n: int
    windowId: str

File: packages/py-parity-checking/src/run-tspice/domains/cells-windows.ts
import type {
  CellsWindowsTargetKind,
  StepCellsWindowsCard,
  StepCellsWindowsInsrtc,
  StepCellsWindowsInsrtd,
  StepCellsWindowsInsrti,
  StepCellsWindowsScard,
  StepCellsWindowsSize,
  StepCellsWindowsSsize,
  StepCellsWindowsValid,
  StepCellsWindowsWncard,
  StepCellsWindowsWnfetd,
  StepCellsWindowsWninsd,
  StepCellsWindowsWnvald,
  StepOutput,
} from "../../case-types.js";
import {
  getOrCreateCharCell,
  getOrCreateDoubleCell,
  getOrCreateIntCell,
  getOrCreateWindow,
  requireCharCell,
  requireDoubleCell,
  requireIntCell,
  requireWindow,
  type RunTspiceContext,
} from "../context.js";

type CellsWindowsStep =
  | StepCellsWindowsCard
  | StepCellsWindowsInsrtc
  | StepCellsWindowsInsrtd
  | StepCellsWindowsInsrti
  | StepCellsWindowsScard
  | StepCellsWindowsSize
  | StepCellsWindowsSsize
  | StepCellsWindowsValid
  | StepCellsWindowsWncard
  | StepCellsWindowsWninsd
  | StepCellsWindowsWnfetd
  | StepCellsWindowsWnvald;

function requireTarget(
  context: RunTspiceContext,
  targetKind: CellsWindowsTargetKind,
  targetId: string,
) {
  switch (targetKind) {
    case "int":
      return requireIntCell(context, targetId);
    case "double":
      return requireDoubleCell(context, targetId);
    case "char":
      return requireCharCell(context, targetId);
    case "window":
      return requireWindow(context, targetId);
    default: {
      const exhaustive: never = targetKind;
      throw new Error(`Unsupported cells-windows targetKind: ${exhaustive}`);
    }
  }
}

/** Execute one `cells-windows.*` workflow step in tspice. */
export function runCellsWindowsStep(
  context: RunTspiceContext,
  step: CellsWindowsStep,
): StepOutput {
  switch (step.op) {
    case "cells-windows.card": {
      const target = requireTarget(context, step.targetKind, step.targetId);
      const value = context.spice.raw.card(target);
      return { op: step.op, value };
    }

    case "cells-windows.insrtc": {
      const cell = getOrCreateCharCell(
        context,
        step.cellId,
        step.maxCardinality ?? 8,
        step.length ?? 32,
      );
      context.spice.raw.insrtc(step.item, cell);
      return { op: step.op, value: null };
    }

    case "cells-windows.insrtd": {
      const cell = getOrCreateDoubleCell(context, step.cellId, step.maxCardinality ?? 8);
      context.spice.raw.insrtd(step.item, cell);
      return { op: step.op, value: null };
    }

    case "cells-windows.insrti": {
      const cell = getOrCreateIntCell(context, step.cellId, step.maxCardinality ?? 8);
      context.spice.raw.insrti(step.item, cell);
      return { op: step.op, value: null };
    }

    case "cells-windows.scard": {
      const target = requireTarget(context, step.targetKind, step.targetId);
      context.spice.raw.scard(step.card, target);
      return { op: step.op, value: null };
    }

    case "cells-windows.size": {
      const target = requireTarget(context, step.targetKind, step.targetId);
      const value = context.spice.raw.size(target);
      return { op: step.op, value };
    }

    case "cells-windows.ssize": {
      const target = requireTarget(context, step.targetKind, step.targetId);
      context.spice.raw.ssize(step.size, target);
      return { op: step.op, value: null };
    }

    case "cells-windows.valid": {
      const target = requireTarget(context, step.targetKind, step.targetId);
      context.spice.raw.valid(step.size, step.n, target);
      return { op: step.op, value: null };
    }

    case "cells-windows.wncard": {
      const window = requireWindow(context, step.windowId);
      const value = context.spice.raw.wncard(window);
      return { op: step.op, value };
    }

    case "cells-windows.wninsd": {
      const window = getOrCreateWindow(context, step.windowId, step.maxIntervals ?? 8);
      context.spice.raw.wninsd(step.left, step.right, window);
      return { op: step.op, value: null };
    }

    case "cells-windows.wnfetd": {
      const window = requireWindow(context, step.windowId);
      const [left, right] = context.spice.raw.wnfetd(window, step.index);
      return { op: step.op, value: { left, right } };
    }

    case "cells-windows.wnvald": {
      const window = requireWindow(context, step.windowId);
      context.spice.raw.wnvald(step.size, step.n, window);
      return { op: step.op, value: null };
    }

    default: {
      const exhaustive: never = step;
      throw new Error(`Unhandled cells-windows step: ${JSON.stringify(exhaustive)}`);
    }
  }
}
